src/util/color/rgbcolor.h

```diff
@@ -0,0 +1,120 @@
+        DEBUG_ASSERT(!isValid());
+    }
+    /*non-explicit*/ RgbColor(RgbColorCode code)
+            : m_color(codeToColor(code)) {
+        DEBUG_ASSERT(m_color == normalizeAnyColor(m_color));
+    }
+    /*non-explicit*/ RgbColor(std::optional<RgbColorCode> optionalCode)
+            : m_color(optionalCodeToColor(optionalCode)) {
+        DEBUG_ASSERT(m_color == normalizeAnyColor(m_color));
+    }
+    // Implicit conversion from QColor without normalization.
+    // Only use this conversion constructor if the argument
+    // matches an RgbColor! Otherwise a debug assertion will
+    // be triggered.
+    /*non-explicit*/ RgbColor(QColor color)
+            : m_color(color) {
+        DEBUG_ASSERT(m_color == normalizeAnyColor(m_color));
+    }
+    // Explicit conversion from any QColor with implicit
+    // normalization that results in a well defined RgbColor.
+    // Use this static function instead of the conversion
+    // constructor to ensure that the resulting RgbColor is
+    // well defined when converting from some QColor.
+    static RgbColor fromAnyColor(QColor anyColor) {
+        return RgbColor(normalizeAnyColor(anyColor));
+    }
+
+    // Implicit conversion into the corresponding QColor.
+    operator QColor() const {
+        return m_color;
+    }
+
+    // Checks if the color is valid or represents "no color",
+    // i.e. is missing or undefined. If the corresponding color
+    // code is stored in a database then the colum value is NULL.
+    bool isValid() const {
+        return m_color.isValid();
+    }
+
+    // Returns the corresponding, optional RGB color code.
+    std::optional<RgbColorCode> optionalCode() const {
+        return colorToOptionalCode(m_color);
+    }
+
+    friend bool operator==(const RgbColor& lhs, const RgbColor& rhs) {
+        return lhs.m_color == rhs.m_color;
+    }
+
+  private:
+    static const RgbColorCode kRgbCodeMask = 0x00FFFFFF;
+    static const RgbColorCode kAlphaCodeMask = 0xFF000000;
+
+    static QColor codeToColor(RgbColorCode code) {
+        return QColor((code & kRgbCodeMask) | kAlphaCodeMask);
+    }
+    static QColor optionalCodeToColor(std::optional<RgbColorCode> optionalCode) {
+        if (optionalCode.has_value()) {
+            return codeToColor(optionalCode.value());
+        } else {
+            return QColor();
+        }
+    }
+
+    static RgbColorCode validColorToCode(QColor color) {
+        DEBUG_ASSERT(color.isValid());
+        return color.rgb() & kRgbCodeMask;
+    }
+    static std::optional<RgbColorCode> colorToOptionalCode(QColor color) {
+        if (color.isValid()) {
+            return std::make_optional(validColorToCode(color));
+        } else {
+            return std::nullopt;
+        }
+    }
+
+    static QColor normalizeAnyColor(QColor anyColor) {
+        if (anyColor.isValid()) {
+            return codeToColor(validColorToCode(anyColor));
+        } else {
+            return QColor();
+        }
+    }
+
+    QColor m_color;
```

Comment:
I mean this: 

RgbColorCode m_color

Reply:
> > > use RgbColorCode as storage variable
> 
> > What does that mean, any links? Please elaborate.
> 
> See my inline comment. You may use 0xFFFFFFFF for invalid. That automatically defaults to transparent, which is what we want in most cases.

Nope. Alpha = 0x00 is transparent.

We agreed on using `QColor` internally and this is the consequent follow-up and reasonable software design for such an approach. `RgbColor` allows even less variations than `QColor`.

We could also store `RgbColorCode` internally with alpha = 0xFF for all valid opaque colors and 0x00000000 as a special value for *invalid*/*transparent*/*no color*. But this custom mapping must not be exposed publicly! The only conversions that are provided are to/from `QColor` and `std::optional<RgbColorCode>`. The public API wouldn't change.

Reply:
Yes, that was the Idea to keep the whole thing performant. 

What is the public API you reffering to?
We have the CO interface that uses the 24 bit color as double, the ES interface that uses tree values for colors, the file tag that uses also a 24 bit value. 
So it looks like QColor is only required to feed the Qt painting interface.

Reply:
> invalid/transparent/no color

Let's be clear with language. These are not identical concepts. There are contexts in which transparent is a valid value such as track color; calling it invalid would be a misnomer.